File: src/day06/main.rs
use itertools::Itertools;
use shared::puzzle_input;
use shared::string_windows::*;

fn detect(input: &str, marker_len: usize) -> usize {
    for (i, s) in input.windows(marker_len).enumerate() {
        if s.chars().counts().len() == marker_len {
            return i + marker_len;
        }
    }
    0
}

pub fn main() {
    let input = puzzle_input!();
    println!("part 1: {}", detect(&input, 4));
    println!("part 2: {}", detect(&input, 14));
}

File: src/day06/main.py
def day6(input, n):
    test = ''
    for i, c in enumerate(input):
        test += c
        if len(test) == n:
            if n == len(set(test)):
                return i+1
            test = test[1:]

def main(input):
    print(f"part 1: {day6(input, 4)}")
    print(f"part 2: {day6(input, 14)}")